### PR TITLE
Deprecation: groupby, resample default dim.

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -3875,9 +3875,7 @@ def reduce(
             Dataset with this object's DataArrays replaced with new DataArrays
             of summarized data and the indicated dimension(s) removed.
         """
-        if dim is ALL_DIMS:
```
Comment (dcherian) [inline xarray/core/dataset.py]: the two changes in this file are minor unrelated cleanups.

File: xarray/core/groupby.py [hunk 2]
```diff
@@ -780,43 +755,6 @@ def reduce_array(ar):
 
         return self.apply(reduce_array, shortcut=shortcut)
 
-    # TODO remove the following class method and DEFAULT_DIMS after the
```
Comment (dcherian) [inline xarray/core/groupby.py]: blindly followed instructions here...

File: doc/whats-new.rst [hunk 3]
```diff
@@ -37,6 +37,9 @@ Breaking changes
 - Most xarray objects now define ``__slots__``. This reduces overall RAM usage by ~22%
   (not counting the underlying numpy buffers); on CPython 3.7/x64, a trivial DataArray
   has gone down from 1.9kB to 1.5kB.
+- The default dimension for :py:meth:`~xarray.Dataset.groupby`, :py:meth:`~xarray.Dataset.resample`,
+  :py:meth:`~xarray.DataArray.groupby` and :py:meth:`~xarray.DataArray.resample` reductions is now the
+  grouping or resampling dimension.
```
Comment (max-sixty) [inline doc/whats-new.rst]: The next `Caveats` section belongs to the previous bullet!

**Comment by max-sixty** [comment]
Thanks a lot @dcherian

File: doc/whats-new.rst (hunk 3 above)
Comment (dcherian) [reply doc/whats-new.rst]: Thanks

**Comment by shoyer** [comment]
Looks good to me!
